rust/src/array.rs

```diff
@@ -18,447 +18,562 @@
-        for i in 0..self.len {
-            let x = *self.data.get(i) + *other.data.get(i);
-            builder.push(x);
+/// A special type of `ListArray` whose elements are binaries.
+pub struct BinaryArray {
+    data: ArrayDataRef,
+    value_offsets: RawPtrBox<i32>,
+    value_data: RawPtrBox<u8>,
+}
+
+impl BinaryArray {
+    /// Returns the element at index `i` as a byte slice.
+    pub fn get_value(&self, mut i: i64) -> &[u8] {
+        i += self.data.offset();
+        unsafe {
+            let pos = self.value_offset_at(i);
+            ::std::slice::from_raw_parts(
+                self.value_data.get().offset(pos as isize),
+                (self.value_offset_at(i + 1) - pos) as usize,
+            )
         }
-        PrimitiveArray::from(builder.finish())
     }
-}
 
-impl<T> Array for PrimitiveArray<T>
-where
-    T: ArrowPrimitiveType,
-{
-    fn len(&self) -> usize {
-        self.len
+    /// Returns the element at index `i` as a string.
+    pub fn get_string(&self, i: i64) -> String {
+        let slice = self.get_value(i);
+        unsafe { String::from_utf8_unchecked(Vec::from(slice)) }
     }
-    fn null_count(&self) -> usize {
-        self.null_count
+
+    /// Returns the offset for the element at index `i`.
+    #[inline]
+    pub fn value_offset(&self, i: i64) -> i32 {
+        self.value_offset_at(i)
     }
-    fn validity_bitmap(&self) -> &Option<Bitmap> {
-        &self.validity_bitmap
+
+    /// Returns the length for the element at index `i`.
+    #[inline]
+    pub fn value_length(&self, mut i: i64) -> i32 {
+        i += self.data.offset();
+        self.value_offset_at(i + 1) - self.value_offset_at(i)
     }
-    fn as_any(&self) -> &Any {
-        self
+
+    #[inline]
+    fn value_offset_at(&self, i: i64) -> i32 {
+        unsafe { *self.value_offsets.get().offset(i as isize) }
     }
 }
 
-/// Create a BufferArray<T> from a Buffer<T> without null values
-impl<T> From<Buffer<T>> for PrimitiveArray<T>
-where
-    T: ArrowPrimitiveType,
-{
-    fn from(data: Buffer<T>) -> Self {
-        PrimitiveArray {
-            len: data.len(),
-            data: data,
-            validity_bitmap: None,
-            null_count: 0,
+impl From<ArrayDataRef> for BinaryArray {
+    fn from(data: ArrayDataRef) -> Self {
+        assert!(data.buffers().len() == 2);
+        let value_offsets = data.buffers()[0].raw_data() as *const u8 as *const i32;
+        let value_data = data.buffers()[1].raw_data();
+        Self {
+            data: data.clone(),
+            value_offsets: RawPtrBox::new(value_offsets),
+            value_data: RawPtrBox::new(value_data),
         }
     }
 }
 
-/// Create a BufferArray<T> from a Vec<T> of primitive values
-impl<T> From<Vec<T>> for PrimitiveArray<T>
-where
-    T: ArrowPrimitiveType + 'static,
-{
-    fn from(vec: Vec<T>) -> Self {
-        PrimitiveArray::from(Buffer::from(vec))
+impl<'a> From<Vec<&'a str>> for BinaryArray {
+    fn from(v: Vec<&'a str>) -> Self {
```

Comment:
As a side note, I suppose you'll want to add a BinaryBuilder at some point?

Reply:
Yes, will add that in a follow-up PR.